pkg/deploy/lattice/service_manager.go

```diff
@@ -2,318 +2,301 @@ package lattice
+		_, ok := newSet[oldSn]
+		if !ok {
+			toDelete = append(toDelete, sn)
+		}
 	}
-	resp, err := latticeSess.ListServiceNetworkServiceAssociationsAsList(ctx, &listServiceNetworkServiceAssociationsInput)
-
-	glog.V(6).Infof("ListServiceNetworkServiceAssociationsAsList req %v, resp %v err %v",
-		listServiceNetworkServiceAssociationsInput, resp, err)
-
-	for _, snAssocResp := range resp {
-		// go through desired SN list
-		needDelete := true
 
-		for _, snName := range snNames {
-			if snName == aws.StringValue(snAssocResp.ServiceNetworkName) {
-				// snName is in the desired SN association list
-				needDelete = false
-			}
+	return toCreate, toDelete, nil
+}
 
-		}
+func svcStatusFromCreateSvcResp(resp *CreateSvcResp) ServiceStatus {
+	status := ServiceStatus{
+		Arn: aws.StringValue(resp.Arn),
+		Id:  aws.StringValue(resp.Id),
+	}
+	if resp.DnsEntry != nil {
+		status.Dns = aws.StringValue(resp.DnsEntry.DomainName)
+	}
+	return status
+}
 
-		if needDelete {
-			svcServiceNetworkInput := vpclattice.DeleteServiceNetworkServiceAssociationInput{
-				ServiceNetworkServiceAssociationIdentifier: snAssocResp.Id,
-			}
+// Create or update Service and ServiceNetwork-Service associations
+func (m *defaultServiceManager) Create(ctx context.Context, svc *Service) (ServiceStatus, error) {
+	emptyStatus := ServiceStatus{}
 
-			svcServiceNetworkOutput, err := latticeSess.DeleteServiceNetworkServiceAssociationWithContext(ctx, &svcServiceNetworkInput)
+	svcSum, err := m.getService(ctx, svc.LatticeName())
+	if err != nil {
+		return emptyStatus, err
+	}
 
-			glog.V(2).Infof("Disassociate service %v from service network %v, resp %v, err %v",
-				snAssocResp.ServiceName, snAssocResp.ServiceNetworkName, svcServiceNetworkOutput, err)
+	var status ServiceStatus
+	if svcSum == nil {
+		status, err = m.createServiceAndSnAssoc(ctx, svc)
+	} else {
+		status, err = m.updateServiceAndSnAssoc(ctx, svc, svcSum)
+	}
+	if err != nil {
+		return emptyStatus, err
+	}
+	return status, nil
+}
 
-			if err != nil {
-				return err
-			}
-			// also check the status, disassociation is still in progress, retry later
-			if aws.StringValue(svcServiceNetworkOutput.Status) == vpclattice.ServiceNetworkServiceAssociationStatusDeleteInProgress {
-				glog.V(6).Infof("Disassociate-in-progress will retry later service %v from service network %v",
-					snAssocResp.ServiceName, snAssocResp.ServiceNetworkName)
-				return errors.New(LATTICE_RETRY)
-			}
+func (m *defaultServiceManager) Delete(ctx context.Context, svc *Service) error {
+	svcSum, err := m.getService(ctx, svc.LatticeName())
+	if err != nil {
+		return err
+	}
+	if svcSum == nil {
+		return nil
+	}
 
-		}
+	svc.Spec.ServiceNetworkNames = []string{}
+	err = m.updateSnSvcAssocs(ctx, svc, svcSum)
+	if err != nil {
+		return err
+	}
 
+	delInput := vpclattice.DeleteServiceInput{
+		ServiceIdentifier: svcSum.Id,
 	}
-	return nil
+	_, err = m.cloud.Lattice().DeleteServiceWithContext(ctx, &delInput)
```

Comment:
At this point, could we be dealing with a service that doesn't have our management tags? Should we be checking? How do you envision upgrades working from a previous version of the controller to this one?

Reply:
There is not tag check on service deletion, and it wasnt before as far as I can tell. About moving forward - while we still at alpha stage and breaking changes happens, there is little concern about backward compatibility. Starting from Beta we cant do that.

Reply:
My initial thought was to make something like "ensureTag" that will go through existing associations and tag them if not tagged. But then controller does not about any associations right now. If I restart controller local storage is gone, and next reconcile event can already mutate list of SN's. So at this point we need to have any tag, that can be a starting point for future upgrades.

Reply:
will do tags in another PR, removed from this